ext/ScreenShot/ScreenShot.h

```diff
@@ -5,72 +5,186 @@
-	{
-		std::move(inData,inData+rowSize,outData);
-		inData += imagePitch;
-		if (flipY)
-			outData -= imagePitch;
-		else
-			outData += imagePitch;
-	}
-
-	asset::IAssetWriter::SAssetWriteParams wparams(img.get());
-	mgr->writeAsset(_outFile, wparams);
-}
-
-template<typename PathOrFile>
-void dirtyCPUStallingScreenshot(IrrlichtDevice* device, const PathOrFile& _outFile, video::IGPUImage* source, uint32_t sourceMipLevel = 0u, bool flipY=true)
-{
-	auto texSize = source->getSize();
-	if (outputFormatOverride==asset::EF_UNKNOWN)
-		outputFormatOverride = source->getColorFormat();
+					auto image = asset::ICPUImage::create(std::move(imgInfo));
+					const auto texelFormatBytesize = getTexelOrBlockBytesize(image->getCreationParameters().format);
+
+					auto texelBuffer = core::make_smart_refctd_ptr<asset::ICPUBuffer>(image->getImageDataSizeInBytes());
+					auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<asset::ICPUImage::SBufferCopy>>(1u);
+					asset::ICPUImage::SBufferCopy& region = regions->front();
+
+					region.imageSubresource.mipLevel = 0u;
+					region.imageSubresource.baseArrayLayer = 0u;
+					region.imageSubresource.layerCount = 1u;
+					region.bufferOffset = 0u;
+					region.bufferRowLength = image->getCreationParameters().extent.width;
+					region.bufferImageHeight = 0u;
+					region.imageOffset = { 0u, 0u, 0u };
+					region.imageExtent = image->getCreationParameters().extent;
+
+					image->setBufferAndRegions(std::move(texelBuffer), regions);
+
+					asset::ICPUImageView::SCreationParams imgViewInfo;
+					imgViewInfo.image = std::move(image);
+					imgViewInfo.format = colorBuffer ? asset::EF_R8G8B8A8_SRGB : asset::EF_D24_UNORM_S8_UINT;
+					imgViewInfo.viewType = asset::IImageView<asset::ICPUImage>::ET_2D;
+					imgViewInfo.flags = static_cast<asset::ICPUImageView::E_CREATE_FLAGS>(0u);
+					imgViewInfo.subresourceRange.baseArrayLayer = 0u;
+					imgViewInfo.subresourceRange.baseMipLevel = 0u;
+					imgViewInfo.subresourceRange.layerCount = imgInfo.arrayLayers;
+					imgViewInfo.subresourceRange.levelCount = imgInfo.mipLevels;
+
+					auto imageView = asset::ICPUImageView::create(std::move(imgViewInfo));
+					auto gpuImageView = driver->getGPUObjectsFromAssets(&imageView.get(), &imageView.get() + 1)->front();
+
+					return std::move(gpuImageView);
+				};
+
+				auto gpuImageViewDepthBuffer = createAttachement(false);
+				auto gpuImageViewColorBuffer = createAttachement(true);
+
+				auto frameBuffer = driver->addFrameBuffer();
+				frameBuffer->attach(video::EFAP_DEPTH_ATTACHMENT, std::move(gpuImageViewDepthBuffer));
+				frameBuffer->attach(video::EFAP_COLOR_ATTACHMENT0, std::move(gpuImageViewColorBuffer));
+
+				return frameBuffer;
+			};
+
+			/*
+				Download mip level image with gpu image usage and save it to IGPUBuffer.
+				Because of the fence placed by driver the function stalls the CPU 
+				to wait on the GPU to finish, beware of that.
```

Comment:
placing the fence does not stall the CPU, you need to wait on the fence with `IDriverFence::waitCPU` before you acess the IGPUBuffer's new data with the CPU, for example, through a mapping

Reply:
I wrote it in previous PR in fence header